_posts/2015-01-30-cloudpipe.markdown

```diff
@@ -0,0 +1,103 @@
+
+ 1. It gave me a concrete spec to target while I was mapping out the endpoints for the API server. This saved us a lot of bikeshedding in the early days.
+ 2. It let me start backend-first and test things with real code, just as a user would experience them. Otherwise, I would have had to use awkward `curl` scripts until we could design and put together a decent SDK.
+ 3. In theory, people are [already using](https://github.com/search?utf8=%E2%9C%93&q=import+multyvac&type=Code&ref=searchresults) the SDK. (If that search is any indication, though, in retrospect we should have stuck with its predecessor, [cloud](https://github.com/search?utf8=%E2%9C%93&q=import+cloud&type=Code&ref=searchresults), instead.)
+
+As a result, the only client SDK that exists today is for Python (and Python 2, at that). That's going to change, though - we've been planning to [write our own client for Python](https://github.com/cloudpipe/cloudpipe/issues/17) and [Shaunak Kashyap](https://github.com/ycombinator) is already working on [an SDK for node.js](https://github.com/cloudpipe/sdk-node). We're also likely to diverge more and more from multyvac as we move on. Stay tuned for more on that front.
+
+Once you have an account running on a cluster somewhere, you can install multyvac with:
+
+```bash
+pip install multyvac
+```
+
+Then, you can submit functions as jobs:
+
+```python
+import multyvac
+
+# Connect to the cluster
+multyvac.config.set_key(api_key='username',
+                        api_secret_key='54321',
+                        api_url='https://somewhere/v1')
+
+def add(a, b):
+  return a + b
+
+# Submit the function as a job. Args and kwargs will be
+# passed to it server-side.
+jid = multyvac.submit(add, 3, 4)
+
+# Use the job ID to get a Job object...
+job = multyvac.get(jid)
+
+# ... and use the Job object to get the result.
+# This call will block until the result is ready.
+result = job.get_result()
+```
+
+Multyvac uses some neat pickling tricks to send your function over the wire: you can actually send closures, or code that references external libraries, and it'll work. Other languages' SDKs may not be able to pull the same range of tricks, depending on how evil each runtime lets us be, but they'll support the same basic idea.
+
+One important thing to notice is that the way work is submitted to the API is completely language-agnostic: it's a contract between the SDK and the base Docker image that's invoked on the server. The Cloudpipe API speaks strictly in terms of POSIX things: command and arguments, environment variables, and standard input, output and error streams.
+
+## How it works
+
+Behind that URL, Cloudpipe is built from a small set of parts, all written in Go.
+
+*[cloudpipe/cloudpipe](https://github.com/cloudpipe/cloudpipe)*
+
+The **API server** is an HTTPS server that hosts a JSON API. It's a fairly thin layer that performs  CRUD operations on job models persisted in a MongoDB instance.
```

Comment:
The good news is, because I'm hiding all of the storage operations behind an interface, we could yank it out or make it configurable without too much trouble :wrench: